src/team_lib/mhba/xccl_mhba_ib.c

```diff
@@ -6,6 +6,312 @@
 #include "xccl_mhba_ib.h"
 #include "utils/utils.h"
 
+static pthread_spinlock_t mmio_spinlock;
+
+static __attribute__((constructor)) void lock_constructor(void)
+{
+    pthread_spin_init(&mmio_spinlock, PTHREAD_PROCESS_PRIVATE);
+}
+
+static inline void mmio_wc_spinlock(pthread_spinlock_t *lock)
+{
+    pthread_spin_lock(lock);
+    //#if !defined(__i386__) && !defined(__x86_64__)
+    //	/* For x86 the serialization within the spin lock is enough to
+    //	 * strongly order WC and other memory types. */
+    //	mmio_wc_start();
+    //#endif
+}
+
+static inline void mmio_wc_spinunlock(pthread_spinlock_t *lock)
+{
+    /* It is possible that on x86 the atomic in the lock is strong enough
+	 * to force-flush the WC buffers quickly, and this SFENCE can be
+	 * omitted too. */
+    mmio_flush_writes();
+    pthread_spin_unlock(lock);
+}
+
+static void pthread_mmio_write64_be(void *addr, __be64 val)
+{
+    __be32 first_dword  = htobe32(be64toh(val) >> 32);
+    __be32 second_dword = htobe32(be64toh(val));
+
+    /* The WC spinlock, by definition, provides global ordering for all UC
+	   and WC stores within the critical region. */
+    mmio_wc_spinlock(&mmio_spinlock);
+
+    mmio_write32_be(addr, first_dword);
+    mmio_write32_be(addr + 4, second_dword);
+
+    mmio_wc_spinunlock(&mmio_spinlock);
+}
+
+#define HAVE_FUNC_ATTRIBUTE_IFUNC 1
+
+#if HAVE_FUNC_ATTRIBUTE_IFUNC
```

Comment:
why need this #if ? you have defined the variable 1 line above - it is always true